### PR TITLE
executor: introduce Executor abstract base class interface

### PR DESCRIPTION
**Pull request by cjp256**
Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_providers/executor.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def create_file(
+        self,
+        *,
+        destination: pathlib.Path,
+        content: bytes,
+        file_mode: str,
+        group: str = "root",
+        user: str = "root",
+    ) -> None:
+        """Create file with content and file mode.
```
Comment (facundobatista) [inline craft_providers/executor.py:38]: "Create *a* file..."?

File: craft_providers/executor.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+        :param destination: Path to file.
+        :param content: Contents of file.
+        :param file_mode: File mode string (e.g. '0644').
+        :param group: File owner group ID.
+        :param user: Filer owner user ID.
```
Comment (facundobatista) [inline craft_providers/executor.py:44]: This contradicts the default above... is it the user/group *id* or the *name*?

File: craft_providers/executor.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    @abstractmethod
+    def pull_file(self, *, source: pathlib.Path, destination: pathlib.Path) -> None:
+        """Copy environment file to host destination.
+
+        Providing this as an abstract method allows the provider to implement
+        the most performant option available.
+
+        :param source: Environment file to copy.
+        :param destination: Host path to copy to.
```
Comment (facundobatista) [inline craft_providers/executor.py:81]: The *destination* must be a directory? Maybe make that explicit in the first docstring line?

File: craft_providers/executor.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        :raises FileNotFound: If source file or destination directory does not
+            exist.
+        :raises OSError: On error copying file.
+        """
+
+    @abstractmethod
+    def push_file(self, *, source: pathlib.Path, destination: pathlib.Path) -> None:
+        """Copy host file into environment at destination.
```
Comment (facundobatista) [inline craft_providers/executor.py:90]: Same than above for destionation being a dir.

(for both) all this implies I can not do `.push_file(source=Path('/tmp/foo'), destination=Path('/tmp/foo')`?

File: docs/index.rst [hunk 5]
```diff
@@ -6,6 +6,7 @@ Here you will find all of the provider documentation...
 .. toctree::
    :caption: Public APIs:
 
+   executors
```
Comment (facundobatista) [inline docs/index.rst:9]: I'm not sure what this line is about.

**Review by facundobatista — COMMENTED** [review]
Nice! Added some questions, thanks!!!

File: docs/index.rst (hunk 5 above)
Comment (cjp256) [reply docs/index.rst:9]: This includes the executors module documentation to the docs.

File: craft_providers/executor.py (hunk 2 above)
Comment (cjp256) [reply craft_providers/executor.py:44]: It could be either, as supported by chown.  Will update docs.

File: craft_providers/executor.py (hunk 3 above)
Comment (cjp256) [reply craft_providers/executor.py:81]: destination would be the path to the file. destination.parent would be the containing directory.  I'll try to word this better.

Comment (facundobatista) [reply craft_providers/executor.py:81]: The confusion comes from a docstring below that says:
```
:raises FileNotFound: If source file or destination directory does not exist.
```

"destination directory"

All this said, if I call `.push_file(source=Path('foo'), destination=Path('/tmp/foo/bar/baz')`, I'm responsible of creating `/tmp/foo/bar/` before? Or the system will take care?

Thanks!

Comment (cjp256) [reply craft_providers/executor.py]: Reworded, hopefully this clarifies :D

File: craft_providers/executor.py (hunk 2 above)
Comment (cjp256) [reply craft_providers/executor.py]: Fixed

File: craft_providers/executor.py (hunk 1 above)
Comment (cjp256) [reply craft_providers/executor.py]: Fixed

**Review by sergiusens — DISMISSED** [review]

**Comment by cjp256** [comment]
Rebased - i forgot squash/rebase options are disabled here.

File: craft_providers/executor.py (hunk 3 above)
Comment (facundobatista) [reply craft_providers/executor.py]: Yes thanks.

File: craft_providers/executor.py (hunk 4 above)
Comment (facundobatista) [reply craft_providers/executor.py]: Yes thanks.

**Review by facundobatista — APPROVED** [review]
Everything looks great now, thanks!

**Pull request merged**